File: Player/VideoCloudBasicPlayer/objc/VideoCloudBasicPlayer/NowPlayingHandler.m
//
//  NowPlayingHandler.m
//  VideoCloudBasicPlayer
//
//  Copyright © 2024 Brightcove, Inc. All rights reserved.
//

#import "NowPlayingHandler.h"


static void * const KVOContext = (void*)&KVOContext;


@interface NowPlayingHandler () <BCOVPlaybackSessionConsumer>

@property (nonatomic, weak) id<BCOVPlaybackSession> session;
@property (nonatomic, strong) NSMutableDictionary *nowPlayingInfo;

@end


@implementation NowPlayingHandler

- (instancetype)initWithPlaybackController:(id<BCOVPlaybackController>)playbackController
{
    if (self = [super init])
    {
        [playbackController addSessionConsumer:self];

        MPRemoteCommandCenter *center = MPRemoteCommandCenter.sharedCommandCenter;

        [center.pauseCommand addTargetWithHandler:^MPRemoteCommandHandlerStatus(MPRemoteCommandEvent * _Nonnull event) {
            [playbackController pause];

            return MPRemoteCommandHandlerStatusSuccess;
        }];

        [center.playCommand addTargetWithHandler:^MPRemoteCommandHandlerStatus(MPRemoteCommandEvent * _Nonnull event) {
            [playbackController play];

            return MPRemoteCommandHandlerStatusSuccess;
        }];

        [center.changePlaybackPositionCommand addTargetWithHandler:^MPRemoteCommandHandlerStatus(MPRemoteCommandEvent * _Nonnull event) {

            MPChangePlaybackPositionCommandEvent *playbackPositionCommandEvent = (MPChangePlaybackPositionCommandEvent *)event;
            CMTime seconds = CMTimeMakeWithSeconds(playbackPositionCommandEvent.positionTime, 600);
            [playbackController seekToTime:seconds completionHandler:nil];

            return MPRemoteCommandHandlerStatusSuccess;
        }];

        __weak typeof(self) weakSelf = self;
        [center.togglePlayPauseCommand addTargetWithHandler:^MPRemoteCommandHandlerStatus(MPRemoteCommandEvent * _Nonnull event) {
            __strong typeof(weakSelf) strongSelf = weakSelf;

            if (strongSelf.session.player.timeControlStatus == AVPlayerTimeControlStatusPaused)
            {
                [playbackController play];
            } else {
                [playbackController pause];
            }

            return MPRemoteCommandHandlerStatusSuccess;
        }];
    }

    return self;
}

- (void)setSession:(id<BCOVPlaybackSession>)session
{
    if (_session)
    {
        [(NSObject *)_session removeObserver:self
                                  forKeyPath:@"player.rate"];
    }

    [(NSObject *)session addObserver:self
                          forKeyPath:@"player.rate"
                             options:NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionNew
                             context:KVOContext];

    _session = session;
}

- (void)updateNowPlayingInfoForAudioOnly
{
    NSDictionary *props = self.session.video.properties;

    self.nowPlayingInfo[MPMediaItemPropertyMediaType] = @(MPMediaTypeMusic);

    // These custom_fields values can be configured in VideoCloud
    // https://beacon.support.brightcove.com/syncing-with-video-cloud/vc-custom-fields.html
    NSDictionary *customFields = props[@"custom_fields"];
    if (customFields[@"album_artist"])
    {
        self.nowPlayingInfo[MPMediaItemPropertyArtist] = customFields[@"album_artist"];
    }

    if (customFields[@"album_name"])
    {
        self.nowPlayingInfo[MPMediaItemPropertyAlbumTitle] = customFields[@"album_name"];
    }

    MPNowPlayingInfoCenter *infoCenter = MPNowPlayingInfoCenter.defaultCenter;
    infoCenter.nowPlayingInfo = self.nowPlayingInfo;
}


#pragma mark - KVO

- (void)observeValueForKeyPath:(NSString *)keyPath
                      ofObject:(id)object
                        change:(NSDictionary<NSKeyValueChangeKey,id> *)change
                       context:(void *)context
{
    if (context != KVOContext)
    {
        [super observeValueForKeyPath:keyPath
                             ofObject:object
                               change:change
                              context:context];
        return;
    }

    if ([object isEqual:self.session] &&
        [keyPath isEqualToString:@"player.rate"])
    {
        MPNowPlayingInfoCenter *infoCenter = MPNowPlayingInfoCenter.defaultCenter;
        NSNumber *rateNumber = change[NSKeyValueChangeNewKey];
        self.nowPlayingInfo[MPNowPlayingInfoPropertyPlaybackRate] = rateNumber;
        infoCenter.nowPlayingInfo = self.nowPlayingInfo;
    }
}


#pragma mark - BCOVPlaybackSessionConsumer

- (void)didAdvanceToPlaybackSession:(id<BCOVPlaybackSession>)session
{
    self.session = session;

    MPNowPlayingInfoCenter *infoCenter = MPNowPlayingInfoCenter.defaultCenter;

    self.nowPlayingInfo = @{}.mutableCopy;
    self.nowPlayingInfo[MPMediaItemPropertyTitle] = [session.video localizedNameForLocale:nil];

    NSNumber *durationNumber = session.video.properties[BCOVVideo.PropertyKeyDuration];
    self.nowPlayingInfo[MPMediaItemPropertyPlaybackDuration] = @(durationNumber.doubleValue / 1000);

    infoCenter.nowPlayingInfo = self.nowPlayingInfo;

    NSString *posterURL = session.video.properties[BCOVVideo.PropertyKeyPoster];
    NSURL *url = [NSURL URLWithString:posterURL];

    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
        NSData *imageData = [NSData dataWithContentsOfURL:url];
        UIImage *image = [UIImage imageWithData:imageData];
        self.nowPlayingInfo[MPMediaItemPropertyArtwork] =
        [[MPMediaItemArtwork alloc] initWithBoundsSize:image.size
                                        requestHandler:^UIImage * _Nonnull(CGSize size) {
            return image;
        }];
        infoCenter.nowPlayingInfo = self.nowPlayingInfo;
    });
}

- (void)playbackSession:(id<BCOVPlaybackSession>)session
          didProgressTo:(NSTimeInterval)progress
{
    MPNowPlayingInfoCenter *infoCenter = MPNowPlayingInfoCenter.defaultCenter;
    self.nowPlayingInfo[MPNowPlayingInfoPropertyElapsedPlaybackTime] = @(progress);
    infoCenter.nowPlayingInfo = self.nowPlayingInfo;
}

@end

File: Player/VideoCloudBasicPlayer/VideoCloudBasicPlayer/NowPlayingHandler.swift
//
//  NowPlayingHandler.swift
//  VideoCloudBasicPlayer
//
//  Copyright © 2026 Brightcove, Inc. All rights reserved.
//

import UIKit
import BrightcovePlayerSDK


final class NowPlayingHandler: NSObject {

    fileprivate weak var session: BCOVPlaybackSession?

    fileprivate var nowPlayingInfo: [String: AnyHashable] = [:]

    private var commandTargets: [(MPRemoteCommand, Any)] = []

    init(with playbackController: BCOVPlaybackController) {
        super.init()

        playbackController.add(self)

        let center = MPRemoteCommandCenter.shared()

        let pauseTarget = center.pauseCommand.addTarget { _ in
            playbackController.pause()
            return .success
        }
        commandTargets.append((center.pauseCommand, pauseTarget))

        let playTarget = center.playCommand.addTarget { _ in
            playbackController.play()
            return .success
        }
        commandTargets.append((center.playCommand, playTarget))

        let changePlaybackPositionTarget = center.changePlaybackPositionCommand.addTarget { command in
            guard let playbackPositionCommandEvent = command as? MPChangePlaybackPositionCommandEvent else {
                return .commandFailed
            }

            let seconds = CMTime(seconds: playbackPositionCommandEvent.positionTime, preferredTimescale: 600)
            playbackController.seek(to: seconds, completionHandler: nil)

            return .success
        }
        commandTargets.append((center.changePlaybackPositionCommand, changePlaybackPositionTarget))

        let togglePlayPauseTarget = center.togglePlayPauseCommand.addTarget { [weak self] _ in
            guard let self, let session, let player = session.player else { return .commandFailed }

            if player.timeControlStatus == .paused {
                playbackController.play()
            } else {
                playbackController.pause()
            }

            return .success
        }
        commandTargets.append((center.togglePlayPauseCommand, togglePlayPauseTarget))
    }

    deinit {
        for (cmd, token) in commandTargets {
            cmd.removeTarget(token)
        }
    }

    func updateNowPlayingInfoForAudioOnly() {
        guard let video = session?.video, let customFields = video.properties["custom_fields"] as? [String: Any] else {
            return
        }

        nowPlayingInfo[MPMediaItemPropertyMediaType] = MPMediaType.music.rawValue

        // These custom_fields values can be configured in VideoCloud
        // https://beacon.support.brightcove.com/syncing-with-video-cloud/vc-custom-fields.html

        if let albumName = customFields["album_name"] as? String {
            nowPlayingInfo[MPMediaItemPropertyAlbumTitle] = albumName
        }

        if let albumArtist = customFields["album_artist"] as? String {
            nowPlayingInfo[MPMediaItemPropertyArtist] = albumArtist
        }

        let infoCenter = MPNowPlayingInfoCenter.default()
        infoCenter.nowPlayingInfo = nowPlayingInfo
    }
}


extension NowPlayingHandler {

    override func observeValue(forKeyPath keyPath: String?,
                               of object: Any?,
                               change: [NSKeyValueChangeKey: Any]?,
                               context: UnsafeMutableRawPointer?) {

        if let session = self.session as? NSObject,
           let object = object as? NSObject,
           session == object,
           keyPath == "player.rate",
           let change,
           let rate = change[.newKey] as? NSNumber {
            let infoCenter = MPNowPlayingInfoCenter.default()
            nowPlayingInfo[MPNowPlayingInfoPropertyPlaybackRate] = rate
            infoCenter.nowPlayingInfo = nowPlayingInfo
        }
    }
}


// MARK: - BCOVPlaybackSessionConsumer

extension NowPlayingHandler: BCOVPlaybackSessionConsumer {

    func didAdvance(to session: BCOVPlaybackSession!) {

        if let prevSession = self.session as? NSObject {
            prevSession.removeObserver(self,
                                       forKeyPath: "player.rate")
        }

        self.session = session

        if let newSession = session as? NSObject {
            newSession.addObserver(self,
                                   forKeyPath: "player.rate",
                                   options: [.new, .initial],
                                   context: nil)
        }

        nowPlayingInfo = [:]

        guard let video = session.video, let videoName = video.localizedName(forLocale: nil),
              let durationNum = video.properties[BCOVVideo.PropertyKeyDuration] as? NSNumber else {
            return
        }

        let duration = Double(durationNum.doubleValue / 1000)

        nowPlayingInfo[MPMediaItemPropertyTitle] = videoName
        nowPlayingInfo[MPMediaItemPropertyPlaybackDuration] = NSNumber(floatLiteral: duration)
        nowPlayingInfo[MPMediaItemPropertyMediaType] = MPMediaType.anyVideo.rawValue

        let infoCenter = MPNowPlayingInfoCenter.default()
        infoCenter.nowPlayingInfo = nowPlayingInfo

        if let posterURL = video.properties[BCOVVideo.PropertyKeyPoster] as? String,
           let url = URL(string: posterURL) {
            URLSession.shared.dataTask(with: url) { [weak self] (data: Data?,
                                                                 response: URLResponse?,
                                                                 error: Error?) in
                if let error {
                    print("Error getting thumbnail image data: \(error.localizedDescription)")
                    return
                }

                DispatchQueue.main.async {
                    guard let self,
                          let data,
                          let image = UIImage(data: data) else {
                        return
                    }

                    self.nowPlayingInfo[MPMediaItemPropertyArtwork] = MPMediaItemArtwork(boundsSize: image.size) { _ in image }

                    let infoCenter = MPNowPlayingInfoCenter.default()
                    infoCenter.nowPlayingInfo = self.nowPlayingInfo
                }
            }.resume()
        }
    }

    func playbackSession(_ session: BCOVPlaybackSession!,
                         didProgressTo progress: TimeInterval) {
        if progress.isInfinite { return }
        let infoCenter = MPNowPlayingInfoCenter.default()
        nowPlayingInfo[MPNowPlayingInfoPropertyElapsedPlaybackTime] = NSNumber(integerLiteral: Int(progress))
        infoCenter.nowPlayingInfo = nowPlayingInfo
    }
}
